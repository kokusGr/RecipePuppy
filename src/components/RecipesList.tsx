import React from 'react'
import { TouchableOpacity, FlatList, View, Image, Text, StyleSheet } from 'react-native'
import { NavigationScreenProp } from 'react-navigation'

import { Recipe } from '../types'

interface Props {
  navigation: NavigationScreenProp<{}>
  recipes: Recipe[]
  onEndReached(): void
}

// NOTE: Some recipes are returned multiple times and no piece of data is unique.
// List items don't change positions so index is good enough in this scenario.
const keyExtractor = (_item: Recipe, index: number) => `${index}`

const RecipesList = ({ recipes, onEndReached, navigation }: Props) => {

  return (
    <FlatList
      data={recipes}
      keyExtractor={keyExtractor}
      onEndReached={onEndReached}
      contentContainerStyle={styles.flatListContent}
      renderItem={({ item }) => (
        <TouchableOpacity
          onPress={() => {
            navigation.navigate('Details', { item })
          }}
        >
          <View style={styles.listItem}>
            <Image
              source={{ uri: item.thumbnail }}
              style={{ height: 56, width: 56, backgroundColor: 'gray' }}
            />
            <View style={styles.textContainer}>
              <Text numberOfLines={3} style={styles.listItemTitle}>{item.title.trim()}</Text>
            </View>
          </View>
        </TouchableOpacity>
      )}
    />
  )
}

const styles = StyleSheet.create({

  flatListContent: {
    paddingHorizontal: 4,
  },
  listItem: {
    height: 88,
    flexDirection: 'row',
    alignItems: 'center',
    marginHorizontal: 8,
    paddingRight: 16,
  },
  listItemTitle: {
    flex: 1,
    fontSize: 15,
  },
  textContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    borderBottomWidth: StyleSheet.hairlineWidth,
    height: 88,
    marginLeft: 24,
  },
})

export default RecipesList
