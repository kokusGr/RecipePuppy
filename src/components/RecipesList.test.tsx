// @ts-nocheck
import React from 'react'
import renderer from 'react-test-renderer'
import { TouchableOpacity } from 'react-native';
import RecipesList from './RecipesList'

const recipes = [{
    title: 'Some title',
    href: 'https://someurl.com',
    ingredients: 'onions, carrot',
    thumbnail: 'https://someimage.com',
}]

it('matches snapshot', () => {
    const tree = renderer.create(<RecipesList recipes={recipes}/>).toJSON()
    expect(tree).toMatchSnapshot()
})

it('navigates to Details with recipe', () => {
    const navigation = {
        navigate: jest.fn(),
    }

    const tree = renderer.create(<RecipesList recipes={recipes} navigation={navigation} />)
    tree.root.findByType(TouchableOpacity).props.onPress()
    expect(navigation.navigate).toHaveBeenCalledTimes(1)
    expect(navigation.navigate).toBeCalledWith('Details', { item: recipes[0] })
})
