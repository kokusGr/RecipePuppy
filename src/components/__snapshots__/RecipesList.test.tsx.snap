// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "paddingHorizontal": 4,
    }
  }
  data={
    Array [
      Object {
        "href": "https://someurl.com",
        "ingredients": "onions, carrot",
        "thumbnail": "https://someimage.com",
        "title": "Some title",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 88,
              "marginHorizontal": 8,
              "paddingRight": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://someimage.com",
              }
            }
            style={
              Object {
                "backgroundColor": "gray",
                "height": 56,
                "width": 56,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomWidth": 0.5,
                "flex": 1,
                "flexDirection": "row",
                "height": 88,
                "marginLeft": 24,
              }
            }
          >
            <Text
              numberOfLines={3}
              style={
                Object {
                  "flex": 1,
                  "fontSize": 15,
                }
              }
            >
              Some title
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
